File: apps/admin/domain/types/category.ts
/**
 * Category Domain Types
 *
 * Type definitions for the category domain following Domain-Driven Design principles.
 *
 * @module domain/types/category
 */

// Core Category Entity
export interface Category {
  readonly id: string;
  readonly store_id: string | null;
  name: string;
  slug: string;
  description: string | null;
  image_url: string | null;
  parent_id: string | null;
  sort_order: number;
  is_active: boolean;
  is_global: boolean;
  gst_percentage: number;
  has_buffer: boolean;
  created_at: string;
  updated_at?: string;
  // Audit fields
  readonly created_by: string | null;
  readonly created_at_branch_id: string | null;
  readonly updated_by: string | null;
  readonly updated_at_branch_id: string | null;
}

// Category Level Enum
export enum CategoryLevel {
  MAIN = 'main',
  SUB = 'sub',
  VARIANT = 'variant',
}

// Category Create DTO
export interface CreateCategoryDTO {
  name: string;
  slug: string;
  description?: string;
  image_url?: string;
  parent_id?: string | null;
  sort_order?: number;
  is_active?: boolean;
  is_global?: boolean;
  store_id?: string;
  gst_percentage?: number;
  has_buffer?: boolean;
}

// Category Update DTO
export interface UpdateCategoryDTO {
  name?: string;
  slug?: string;
  description?: string;
  image_url?: string;
  parent_id?: string | null;
  sort_order?: number;
  is_active?: boolean;
  is_global?: boolean;
  store_id?: string;
  gst_percentage?: number;
}

// GST Options for category dropdown
export const GST_OPTIONS = [
  { value: 5, label: '5%' },
  { value: 12, label: '12%' },
  { value: 18, label: '18%' },
] as const;

// Category with Relations
export interface CategoryWithRelations extends Category {
  parent?: Category | null;
  children?: Category[];
  product_count?: number;
  level: CategoryLevel;
  path: string; // e.g., "Costumes > Earrings > Diamond"
}

// Category Hierarchy
export interface CategoryHierarchy {
  mains: Category[];
  subs: Category[];
  variants: Category[];
}

// Category Tree Node
export interface CategoryTreeNode {
  category: Category;
  children: CategoryTreeNode[];
  level: number;
  isExpanded: boolean;
  productCount: number;
}

// Category Validation
export interface CategoryValidationError {
  field: string;
  message: string;
  code: string;
}

export interface CategoryValidationResult {
  is_valid: boolean;
  errors: CategoryValidationError[];
  warnings: CategoryValidationError[];
}

// Type Guards
export const isValidCategory = (obj: any): obj is Category => {
  return obj && 
         typeof obj.id === 'string' &&
         typeof obj.name === 'string' &&
         typeof obj.slug === 'string';
};

export const isMainCategory = (category: Category): boolean => {
  return !category.parent_id;
};

export const isSubCategory = (category: Category, allCategories: Category[]): boolean => {
  if (!category.parent_id) return false;
  const parent = allCategories.find(c => c.id === category.parent_id);
  return parent ? !parent.parent_id : false;
};

export const isVariantCategory = (category: Category, allCategories: Category[]): boolean => {
  if (!category.parent_id) return false;
  const parent = allCategories.find(c => c.id === category.parent_id);
  return parent ? !!parent.parent_id : false;
};

File: apps/mobile/lib/features/categories/models/category.dart
import 'package:equatable/equatable.dart';

/// Category model matching the admin panel's domain/types/category.ts
///
/// Core entity fields: id, name, slug, description, image_url,
/// parent_id (for hierarchy: Main → Sub → Variant), sort_order,
/// is_active, is_global, created_at, updated_at.
class Category extends Equatable {
  final String id;
  final String name;
  final String slug;
  final String? description;
  final String? imageUrl;
  final String? parentId;
  final int sortOrder;
  final bool isActive;
  final bool isGlobal;
  final String createdAt;
  final String? updatedAt;

  // Relations (optional, populated by detail endpoints)
  final Category? parent;
  final List<Category>? children;
  final int? productCount;

  const Category({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    this.imageUrl,
    this.parentId,
    this.sortOrder = 0,
    this.isActive = true,
    this.isGlobal = false,
    required this.createdAt,
    this.updatedAt,
    this.parent,
    this.children,
    this.productCount,
  });

  /// Whether this is a Main (root) category.
  bool get isMain => parentId == null;

  @override
  List<Object?> get props => [
        id,
        name,
        slug,
        description,
        imageUrl,
        parentId,
        sortOrder,
        isActive,
        isGlobal,
        createdAt,
        updatedAt,
        parent,
        children,
        productCount,
      ];

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String? ?? '',
      description: json['description'] as String?,
      imageUrl: json['image_url'] as String?,
      parentId: json['parent_id'] as String?,
      sortOrder: (json['sort_order'] as num?)?.toInt() ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      isGlobal: json['is_global'] as bool? ?? true,
      createdAt: json['created_at'] as String? ?? '',
      updatedAt: json['updated_at'] as String?,
      parent: json['parent'] != null && json['parent'] is Map
          ? Category.fromJson(Map<String, dynamic>.from(json['parent']))
          : null,
      children: json['children'] != null && json['children'] is List
          ? (json['children'] as List)
              .map((c) => Category.fromJson(Map<String, dynamic>.from(c)))
              .toList()
          : null,
      productCount: (json['product_count'] as num?)?.toInt(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'description': description,
      'image_url': imageUrl,
      'parent_id': parentId,
      'sort_order': sortOrder,
      'is_active': isActive,
      'is_global': isGlobal,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
